kafka_consumer_freshness_tracker/src/main/java/com/tesla/data/consumer/freshness/ConsumerFreshness.java

```diff
@@ -169,95 +170,151 @@ void run() {
+   *  - burrow group status lookup fails
+   *  - execution is interrupted
+   * Failure to actually measure the consumer is swallowed into a log message & metric update; obviously, this is less
+   * than ideal for many cases, but it will be addressed later.
+   *
+   * @param burrow access to burrow
+   * @param workers pool of workers available for querying Kafka
+   * @param consumerGroup name of the consumer group to measure
+   * @return a future representing the measurement of each topic-partition the consumer reads. If any partition
+   * failed to be measured the entire future (i.e. calls to {@link Future#get()}) will be considered a failure.
+   * @throws InterruptedException if the application is interrupted while waiting for an available worker
+   */
+  private ListenableFuture<List<Object>> measureConsumer(Burrow.ClusterClient burrow,
+                                                         ArrayBlockingQueue<KafkaConsumer> workers,
+                                                         String consumerGroup) throws InterruptedException {
+    Map<String, Object> status;
+    try {
+      status = burrow.getConsumerGroupStatus(consumerGroup);
+      Preconditions.checkState(status.get("partitions") != null,
+          "Burrow response is missing partitions, got {}", status);
+    } catch (IOException | IllegalStateException e) {
+      // this happens sometimes, when burrow is acting up (e.g. "bad" consumer names)
+      LOG.error("Failed to read Burrow status for consumer {}. Skipping", consumerGroup, e);
+      metrics.error.labels(burrow.getCluster(), consumerGroup).inc();
+      return Futures.immediateFuture(Collections.emptyList());
+    }
+
+    boolean anyEndOffsetFound = false;
+    List<Map<String, Object>> partitions = (List<Map<String, Object>>) status.get("partitions");
+    List<ListenableFuture<Object>> partitionFreshnessComputation = new ArrayList<>(partitions.size());
+    for (Map<String, Object> state : partitions) {
+      String topic = (String) state.get("topic");
+      int partition = (int) state.get("partition");
+      Map<String, Object> end = (Map<String, Object>) state.get("end");
+      if (end == null) {
+        LOG.debug("Skipping {}:{} - {}:{} because no offset found",
+            burrow.getCluster(), consumerGroup, topic, partition);
+        this.metrics.missing.inc();
+        continue;
+      }
+      anyEndOffsetFound = true;
+      long offset = Long.valueOf(end.get("offset").toString());
+      boolean upToDate = Long.valueOf(state.get("current_lag").toString()) == 0;
+      FreshnessTracker.ConsumerOffset consumerState =
+          new FreshnessTracker.ConsumerOffset(burrow.getCluster(), consumerGroup, topic, partition, offset,
+              upToDate);
+
+      // wait for a consumer to become available
+      KafkaConsumer consumer = workers.take();
+      ListenableFuture result = this.executor.submit(new FreshnessTracker(consumerState, consumer, metrics));
```

Comment:
why is the type missing for ListenableFuture

Reply:
Mumble mumble generics mumble mumble. Looks like works fine now though!